# 主題：評估和清理英國電商公司銷售數據

## 分析目標

此數據分析的目的是根據市場銷售數據，找出熱銷產品，以便制定更有效的市場策略來提升營收。

## 簡介

原始數據集記錄了一家英國線上零售公司在2010年12月1日至2011年12月9日期間的所有交易情況，涵蓋了該公司在全球不同國家和地區的業務數據。該公司主要銷售各種情境下的禮品，包括但不限於生日禮品、結婚紀念品、聖誕禮品等等。該公司的客戶群體主要包括批發商和個人消費者，其中批發商佔據了相當大的比例。

數據每列的含義如下：
- `InvoiceNo`: 發票號碼。6位數，作為交易的唯一標識。如果這個代碼以字母 "C" 開頭，表示這筆交易被取消。
- `StockCode`: 產品代碼。5位數，作為產品的唯一標識。
- `Description`: 產品名稱。
- `Quantity`: 產品在交易中的數量。
- `InvoiceDate`: 發票日期和時間。交易發生的日期和時間。
- `UnitPrice`: 單價。價格單位為英鎊（£）。
- `CustomerID`: 客戶編號。5位數，作為客戶的唯一標識。
- `Country`: 國家名稱。客戶所居住的國家的名稱。

## 讀取數據

In [9]:
import pandas as pd
import numpy as np
original_data = pd.read_csv('e_commerce.csv')
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 評估數據整齊度

In [11]:
original_data.sample(10) #先隨意抽樣10行


InvoiceNo StockCode                         Description  Quantity  \
313565    564516     22131    FOOD CONTAINER SET 3 LOVE HEART          1   
431474    573601     23311         VINTAGE CHRISTMAS STOCKING          3   
134922    547872     22328  ROUND SNACK BOXES SET OF 4 FRUITS          1   
65733     541696     37446  MINI CAKE STAND WITH HANGING CAKES         1   
348455    567429     47566                       PARTY BUNTING         4   
52429    C540788     20829    GLITTER HANGING BUTTERFLY STRING        -2   
200032    554112    90082A    DIAMANTE BOW BROOCH GREEN COLOUR         1   
458719    575848     21791  VINTAGE HEADS AND TAILS CARD GAME          5   
240645    558105     21034        REX CASH+CARRY JUMBO SHOPPER         4   
470612    576618     21285            RETROSPOT CANDLE  MEDIUM         1   

             InvoiceDate  UnitPrice  CustomerID         Country  
313565   8/25/2011 14:45       4.13         NaN  United Kingdom  
431474  10/31/2011 15:24       4.96         NaN  United Kingdom  
134922   3/27/2011 14:34       2.95     16777.0  United Kingdom  
65733    1/20/2011 18:08       3.29         NaN  United Kingdom  
348455   9/20/2011 11:39       4.95     17686.0  United Kingdom  
52429    1/11/2011 11:46       2.10     13576.0  United Kingdom  
200032   5/22/2011 14:44       6.35     17611.0  United Kingdom  
458719  11/11/2011 12:10       1.25     14178.0  United Kingdom  
240645   6/26/2011 14:54       0.95     15644.0  United Kingdom  
470612  11/15/2011 17:00       4.13         NaN  United Kingdom

#### 從抽樣的10條數據來看，數據符合每個Column是一個變量，每個行是一個觀察值，每個單元格是一個值。具體來說每行是關於一個商品的一次性交易，每列則是交易相關的各個變量，因此不存在結構性問題。

## 評估數據的乾淨程度

In [14]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


#### 從輸出結果來看，數據結果共有541909條觀察值，而 ```Description``` 和 ```CustomerID``` 存在缺失值。
#### 此外，```InvoiceDate``` 的數據類型應該為日期格式，```CustomerID``` 應為字串類型，這兩個要做數據格式的轉換。

### 評估缺失數據

#### 在了解 ```Description``` 存在缺失值後，根據條件取出缺失的觀察值

In [18]:
# original_data['Description'].isnull() #只返回布林值的Series
original_data[original_data['Description'].isnull()] #將布林條件應用於 DataFrame 來篩選行

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

#### 有1454條交易數據缺失 ```Description``` 變量值
#### 從缺失的結果來看，這些缺失的變量值 ```UnitPrice``` 都為0，為了驗證這個想法，看是否存在 ```Description``` 變量缺失且 ```UnitPrice``` 不為0 的數據。

In [20]:
original_data[(original_data['Description'].isnull()) & (original_data['UnitPrice'] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

#### 篩選出來的數據都是0，驗證想法是正確的。```Description``` 表示產品名稱，```UnitPrice```表示產品單價，都是進行後續產品交易分析的重要變量，如果他們同時缺失，代表這些數據無法提供有效意義，因此這些數據可以在之後被刪除。

#### ```CustomerID``` 變量同時存在缺失值，因此也根據條件提取缺失觀察值

In [23]:
original_data[original_data['CustomerID'].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

#### ```CustomerID``` 表示客戶產品編號，不是分析商品的必要變量，有些 ```CustomerID``` 缺失的交易數據仍然有效，因此保留變量為空的缺失值

### 評估重複數據

#### 雖然 ```InvoiceNo``` 和 ```CustomerID``` 和 ```StockCode``` 都是唯一值，但一次交易可以包含多種商品，因此```InvoiceNo``` 可以存在重複，不同交易也可包含不同商品，因此 ```StockCode``` 可以重複， 顧客也可以進行多次交易和下單多種商品，因此```CustomerID``` 也可以重複。 針對以上的結果，不需要評估重複的數據資料。

### 評估不一致的數據

#### 不一緻數據可能存在於 ```Country``` 的變量中，我們接下來要查看是否有使用不同名稱來描述同一個國家。

In [29]:
original_data['Country'].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


#### 從 ```Country``` 變量值來看，```USA``` 和 ```United States``` 均表示美國，```U.K.``` 和 ```UK``` 和 ```United Kingdom``` 均表示英國，因此要再下一步清洗步驟將這些值變成統一的名稱。

### 評估錯誤和無效數據

#### 透過 ```description``` 方法對數據的統計資訊進行基本瞭解，

In [33]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

#### 從輸出的結果來看，```Quantity``` & ```UnitPrice```有存在負數，會對後續數值分析產生影響，因此我們先篩選出```Quantity```為負數的觀察值。

In [35]:
original_data[original_data['Quantity'] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

#### 從以上結果可以看出 ```Quantity``` 為負值的，而 ```invoiceNo``` 前面開頭都有一個字母Ｃ，為了驗證這個想法，在下一步做驗證。

In [37]:
original_data[(original_data['Quantity'] < 0) & (original_data['InvoiceNo'].str[0] != 'C' )]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

#### 以上驗證錯誤，因為從```InvoiceNo```裡可以看到有些並沒有Ｃ開頭，但在這些數據裡面又看到 ```UnitPrice``` 等於0，因此在篩選條件裡面多加入一個條件，驗證以上的想法是否正確。

In [39]:
original_data[(original_data['Quantity'] < 0) & (original_data['InvoiceNo'].str[0] != 'C' ) & (original_data['UnitPrice'] != 0 )]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

#### 根據輸出結果，當 ```Quantity```為負數的時候，```Invoice```開頭可能為Ｃ或是 ```UnitePrice``` 可能為0，由此可知，這樣的結果對分析暢銷品沒有實質的幫助，因此在數據清洗步驟，會將```Quantity```為負數的數據刪掉。

#### 接下來篩選出 ```UnitPrice``` 也是負數的行

In [42]:
original_data[original_data['UnitPrice'] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

#### 從輸出結果來看 ```UnitPrice``` 的 ```Description``` 指得是壞帳調整，不是實際商品交易數據，因此也需要刪除。

## 清洗數據

#### 根據前面結果顯示，我們要進行以下的數據清洗
- 修改 ```InvoiceDate``` 的格式修改為字串類型
- 修改 ```CustomerID``` 的格式修改為字串類型
- 將 ```Description``` 缺失的變量刪除
- 將 ```USA``` 修改為 ```United States```
- 將 ```U.K.``` 和 ```UK``` 修改為 ```United Kingdom```
- 將 ```Quantity``` 為負數的值刪掉
- 將 ```UnitPrice``` 為負數的值刪掉

#### 為了區隔原始數據和清洗後的數據，創建一個新的變數 ```cleaned_data```，之後清洗完的數據都會被運用在 ```cleaned_data```上面

In [47]:
cleaned_data = original_data.copy()
cleaned_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

### 把```InvoiceDate```轉換成 ```datetime```類型

In [49]:
cleaned_data['InvoiceDate'] = pd.to_datetime(cleaned_data['InvoiceDate'])
cleaned_data['InvoiceDate']
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


### 把```CustomerID```轉換成 字串類型

In [51]:
cleaned_data['CustomerID'] = cleaned_data['CustomerID'].astype(str)
cleaned_data['CustomerID']

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

### 把 ```CustomerID``` 尾巴為.0的字刪除

In [53]:
cleaned_data['CustomerID'].str.slice(0,-2)

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

### 把```Description``` 有缺失的值刪除

In [55]:
cleaned_data.dropna(subset = ['Description'], inplace = True)

In [56]:
cleaned_data['Description'].isnull().sum()

0

### 將 ```USA``` 修改為 ```United States```

In [91]:
cleaned_data['Country'] = cleaned_data['Country'].replace('USA','United States')
len(cleaned_data[cleaned_data['Country'] == 'USA'])

0

### 將 ```U.K.``` 和 ```UK``` 修改為 ```United Kingdom```

In [100]:
cleaned_data['Country'] = cleaned_data['Country'].replace(['U.K.','UK'],'United States')
print(len(cleaned_data[cleaned_data['Country'] == 'U.K.']))
print(len(cleaned_data[cleaned_data['Country'] == 'UK']))

0
0


### 將 ```Quantity``` 為負數的值刪掉

In [112]:
cleaned_data = cleaned_data[cleaned_data['Quantity'] >= 0]
len(cleaned_data[cleaned_data['Quantity'] < 0])

0

### 將 ```UnitPrice``` 為負數的值刪掉

In [118]:
cleaned_data = cleaned_data[cleaned_data['UnitPrice'] >= 0]
len(cleaned_data[cleaned_data['UnitPrice'] < 0])

0

## 保存清理後的數據

#### 完成清理後數據的資料，將文件名命為 ```e_commerce_cleaned``` 

In [140]:
cleaned_data.to_csv('e_commerce_cleaned.csv', index = False)